mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -220,145 +166,301 @@ struct BackpropOpPattern : public OpConversionPattern<BackpropOp> {
+    {
+        MLIRContext *context = moduleOp.getContext();
+        Type indexType = getTypeConverter()->getIndexType();
+        OpBuilder::InsertionGuard insertGuard(builder);
+        builder.setInsertionPointToStart(moduleOp.getBody());
+
+        auto allocationLike = moduleOp.lookupSymbol<LLVM::GlobalOp>(enzyme_allocation_key);
+        if (allocationLike) {
+            return allocationLike;
+        }
+
+        auto ptrType = LLVM::LLVMPointerType::get(context);
+        auto resultType = LLVM::LLVMArrayType::get(ptrType, 4);
+
+        builder.create<LLVM::GlobalOp>(loc, LLVM::LLVMArrayType::get(builder.getI8Type(), 3), true,
+                                       LLVM::Linkage::Linkonce, "dealloc_indices",
+                                       builder.getStringAttr(StringRef("-1", 3)));
+        allocationLike =
+            builder.create<LLVM::GlobalOp>(loc, resultType,
+                                           /*isConstant=*/false, LLVM::Linkage::External,
+                                           enzyme_allocation_key, /*address space=*/nullptr);
+        builder.createBlock(&allocationLike.getInitializerRegion());
+        Value allocFn = builder.create<LLVM::AddressOfOp>(loc, ptrType, allocFuncName);
+        Value sizeArgIndex =
+            builder.create<LLVM::ConstantOp>(loc, indexType, builder.getIndexAttr(0));
+        Value sizeArgIndexPtr = builder.create<LLVM::IntToPtrOp>(loc, ptrType, sizeArgIndex);
+        Value deallocIndicesPtr =
+            builder.create<LLVM::AddressOfOp>(loc, ptrType, "dealloc_indices");
+        Value freeFn = builder.create<LLVM::AddressOfOp>(loc, ptrType, freeFuncName);
+
+        Value result = builder.create<LLVM::UndefOp>(loc, resultType);
+        result = builder.create<LLVM::InsertValueOp>(loc, result, allocFn, 0);
+        result = builder.create<LLVM::InsertValueOp>(loc, result, sizeArgIndexPtr, 1);
+        result = builder.create<LLVM::InsertValueOp>(loc, result, deallocIndicesPtr, 2);
+        result = builder.create<LLVM::InsertValueOp>(loc, result, freeFn, 3);
+        builder.create<LLVM::ReturnOp>(loc, result);
+
+        return allocationLike;
+    }
+
+    static void insertFunctionName(PatternRewriter &rewriter, Operation *op, StringRef key,
+                                   StringRef value)
+    {
+        ModuleOp moduleOp = op->getParentOfType<ModuleOp>();
+        PatternRewriter::InsertionGuard insertGuard(rewriter);
+        rewriter.setInsertionPointToStart(moduleOp.getBody());
+        LLVM::GlobalOp glb = moduleOp.lookupSymbol<LLVM::GlobalOp>(key);
+        if (!glb) {
+            glb = rewriter.create<LLVM::GlobalOp>(
+                moduleOp.getLoc(),
+                LLVM::LLVMArrayType::get(IntegerType::get(rewriter.getContext(), 8), value.size()),
+                true, LLVM::Linkage::Linkonce, key, rewriter.getStringAttr(value));
+        }
+    }
+
+    static LLVM::GlobalOp insertEnzymeFunctionLike(PatternRewriter &rewriter, Operation *op,
```

Comment:
A docstring would also be useful here, in particular to distinguish it from `insertEnzymeAllocationLike`.